---
output: github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# transformGamPoi

<!-- badges: start -->
<!-- badges: end -->

R package that accompanies our paper 'Transformation and Preprocessing of Single-Cell RNA-Seq Data' (https://www.biorxiv.org/content/10.1101/2021.06.24.449781v1).

`transformGamPoi` provides methods to stabilize the variance of single cell count data:

* acosh transformation based on the delta method
* shifted logarithm (log(x + c)) with a pseudo-count c, so that it approximates the acosh transformation
* randomized quantile and Pearson residuals 

## Installation

You can install the current development version of `transformGamPoi` using:
``` r
# install.packages("devtools")
devtools::install_github("const-ae/transformGamPoi")
```
## Example

Let's compare the different variance-stabilizing transformations.

We start by loading the `transformGamPoi` package and setting a seed to make sure the results are reproducible.
```{r loadLibraries}
library(transformGamPoi)
set.seed(1)
```

We then load some example data, which we subset to 1000 genes and 500 cells
```{r loadData}
sce <- TENxPBMCData::TENxPBMCData("pbmc4k")
sce_red <- sce[sample(which(rowSums2(counts(sce)) > 0), 1000),
               sample(ncol(sce), 500)]
```

We calculate the different variance-stabilizing transformations. We can either use the generic `transformGamPoi()` method and specify the `transformation`, or we use the low-level functions `acosh_transform()`, `shifted_log_transform()`, and `residual_transform()` which provide more settings. All functions return a matrix, which we can for example insert back into the `SingleCellExperiment` object:
```{r applyVSTs}
assay(sce_red, "acosh") <- transformGamPoi(sce_red, transformation = "acosh")
assay(sce_red, "shifted_log") <- shifted_log_transform(sce_red, overdispersion = 0.1)
# For large datasets, we can also do the processing without 
# loading the full dataset into memory (on_disk = TRUE)
assay(sce_red, "rand_quant") <- residual_transform(sce_red, "randomized_quantile", on_disk = FALSE)
assay(sce_red, "pearson") <- residual_transform(sce_red, "pearson", clipping = TRUE, on_disk = FALSE)
```

Finally, we compare the variance of the genes after transformation using a scatter plot
```{r plotMeanVar, warning=FALSE}
par(pch = 20, cex = 1.15)
mus <- rowMeans2(counts(sce_red))
plot(mus, rowVars(assay(sce_red, "acosh")), log = "x", col = "black",
     xlab =  "Log Gene Means", ylab = "Variance after transformation")
points(mus, rowVars(assay(sce_red, "shifted_log")), col = "red")
points(mus, rowVars(assay(sce_red, "pearson")), col = "green")
points(mus, rowVars(assay(sce_red, "rand_quant")), col = "blue")
legend("topleft", legend = c("acosh", "shifted log", "Rand. Quantile Resid.", "Pearson Resid."),
       col = c("black", "red", "blue", "green"), pch = 16)
```

# See also

There are a number of preprocessing methods and packages out there. Of particular interests are

* [sctransform](https://github.com/ChristophH/sctransform) by Christoph Hafemeister and the [Satija lab](https://satijalab.org/). The original developers of the Pearson residual variance-stabilizing transformation approach for single cell data.
* [scuttle::logNormCounts()](https://bioconductor.org/packages/release/bioc/html/scuttle.html) by Aaron Lun. This is an alternative to the `shifted_log_transform()` and plays nicely together with the Bioconductor universe. For more information, I highly recommend to take a look at the [normalization](https://bioconductor.org/books/release/OSCA/normalization.html) section of the [OSCA book](https://bioconductor.org/books/release/OSCA/).
* [Sanity](https://github.com/jmbreda/Sanity) by Jérémie Breda _et al._. This method is not directly concerned with variance stabilization, but still provides an interesting approach for single cell data preprocessing.



# Session Info

```{r}
sessionInfo()
```


